jobs/build/incremental/Jenkinsfile

```diff
@@ -0,0 +1,329 @@
+                        $class: 'BooleanParameterDefinition',
+                        defaultValue: false
+                    ],
+                    [
+                        name: 'MOCK',
+                        description: 'Pick up changed job parameters and then exit',
+                        $class: 'BooleanParameterDefinition',
+                        defaultValue: false
+                    ],
+                ]
+            ]
+        ]
+    )
+
+    // User can have the job end if this is just a run to pick up parameter changes
+    // (which Jenkins discovers by running the job).
+    if (MOCK.toBoolean()) {
+        currentBuild.description = "Ran in mock mode"
+        return
+    }
+
+    OSE_MAJOR = MINOR_VERSION.tokenize('.')[0].toInteger() // Store the "X" in X.Y
+    OSE_MINOR = MINOR_VERSION.tokenize('.')[1].toInteger() // Store the "Y" in X.Y
+    currentBuild.displayName = "#${currentBuild.number} - ${OSE_MAJOR}.${OSE_MINOR}"
+
+    EXCLUSIONS = clean_comma_list(EXCLUSIONS)
+    INCLUSIONS = clean_comma_list(INCLUSIONS)
+    def filter = INCLUSIONS ? "-i ${INCLUSIONS}" : "-i ''"
+    if (EXCLUSIONS) {
+        filter = "-x ${EXCLUSIONS}"
+    }
+
+    set_workspace()
+
+    def buildlib = load("pipeline-scripts/buildlib.groovy")
+    buildlib.initialize()
+
+    // doozer_working must be in WORKSPACE in order to have artifacts archived
+    DOOZER_WORKING = "${WORKSPACE}/doozer_working"
+    // Clear out previous work
+    sh "rm -rf ${DOOZER_WORKING}"
+    sh "mkdir -p ${DOOZER_WORKING}"
+    def doozer_opts = "--working-dir ${DOOZER_WORKING} --group openshift-${OSE_MAJOR}.${OSE_MINOR}"
+
+    List<String> images = []
+
+    /* ************************************************************************** */
+    currentBuild.description = "Scanning images"
+    stage('Scan images') {
+
+        try {
+            def yamlData = readYaml text: buildlib.doozer(
+                "${doozer_opts} ${filter} config:scan-sources --yaml", [capture: true]
+            )
+            // gotta parse out the list...
+            yamlData["images"].each {
+                if (it["changed"]) {
+                   images.add(it["name"])
+                }
+            }
+            if (!images) {
+                currentBuild.description = "No images found to update."
+                return
+            }
+            echo "Images that have changed:\n" + images.join("\n")
+
+            def short_images = images.collect()
+            if (images.size() > 2) {
+                short_images = short_images[0..1]
+                short_images.add("...")
+            }
+            currentBuild.description = "Changes to: " + short_images.join(", ")
+
+            // also determine child images
+            yamlData = readYaml text: buildlib.doozer(
+                "${doozer_opts} images:show-tree --yml", [capture: true]
+            )
+            Closure gather_children
+            gather_children = { all, yaml, changed, gather ->
```

Comment:
I wouldn't say I'm (or any ARTist) especially 1337 with groovy right now. Please add some flavor text in & around this.

Reply:
Can do... is it not evident that it's a recursive closure for traversing the child tree?

Reply:
Uh. Kinda.

Reply:
Better?

Reply:
fantastic